[#34] 공연 정보 추가 기능 구현

hanwix2 opened the pull request:
## 개요

*[관리자 권한]*

1. **공연 정보 등록**
    - 공연 정보: 
        - title(공연명), imageFilePath(공연이미지 파일 위치), detail(상세설명), ageLimit(관람가능연령), showType(공연타입), venueId(공연장ID), hallId(공연홀ID)
        - 최초 공연 정보 등록 시 이미지 파일은 Null 값으로 등록(공연 등록 이후 이미지 등록 가능)
        > - 공연 타입(ShowType): 뮤지컬, 연극, 콘서트 
    - *공연장ID, 공연홀ID, 공연타입은 UI에서 선택하여 유효한 값 만을 넣는다고 가정*

2. **포스터 이미지 등록**
    - 기존에 존재하는 공연에 대하여 포스터 이미지 등록
    - *공연ID는 UI에서 선택하여 유효한 값 만을 넣는다고 가정*

3. **공연 스케줄 등록**
    - 기존에 존재하는 공연에 대하여 스케줄 등록
    - 하나의 공연에 여러 개의 스케줄 등록이 가능
    - 스케줄에는 공연 시작 시간과 끝나는 시간을 입력
    - 공연 시간은 최대 24시간으로 제한
    - *공연ID는 UI에서 선택하여 유효한 값 만을 넣는다고 가정*

4. **공연 스케줄 등록 시 시간 중첩 방지**
    - 같은 공연에 여러 스케줄 입력 중 발생 가능한 시간 중첩 방지
    - 문제 상황: 
        1. 끝나는 시간이 시작 시간 보다 빠른 경우
        2. 새로 등록하는 스케줄 간 시간이 중첩되는 경우
        3. 기존 DB에 등록된 스케줄과 시간이 중첩되는 경우


## Progress
- [x] 공연 정보 등록 기능
- [x] 포스터 이미지 등록 기능
- [x] 공연 스케줄 입력 기능
- [x] 같은 홀의 공연 등록 시 스케줄 중복 방지
- [x] 테스트


## Detail
1. **공연 정보 등록 기능**
    - Performance(공연) 관련 Controller, Service, Mapper 생성
        - `insertPerformance `메소드에서 공연 등록 로직 수행
    - `PerformanceRequest `객체로 공연 정보 받아옴
    - 공연 등록 전 공연명과 공연유형이 중복되었으면 `PerformanceAlreadyExistsException `에러
    - 공연 유형은 `ShowType `Enum 및 `TypeHandler` 로 처리

2. **포스터 이미지 등록 기능**
    - **MultipartFile**을 이용하여 파일 업로드 기능 구현
    - PUT 메소드 (공연 정보의 이미지 정보 업데이트 API)
        - 공연 등록 시 Request Body에 MultipartFile과 JSON 형식의 공연 정보를 함께 보낼 수 없으므로 기존 공연 정보 등록 API와 분리하여 구현
    - 이미지 저장 시 기존에 존재하는 파일이 있다면 삭제 후 저장되고 DB에선 새로운 파일의 저장경로로 속성을 update
    - 이미지는 프로젝트 파일의 files/performanceImages 폴더(상대경로)에 저장
    - FileService, FileUtil 클래스를 구현하여 파일 관련 기능 로직을 PerformanceService에서 분리
    - 관련 예외 상황 처리:
        - 지정된 파일 형식만 허용(이미지 파일: jpeg, png, gif)
        - 최대 파일 허용 크기: 10MB

3. **공연 스케줄 등록 기능**
    - **PerformanceTimeRequest** 객체로 공연 시작, 끝 시간 전달 받음
    - 공연 스케줄은 하나 이상이므로 PerformanceTimeRequest는 List로 받음
    - 이후 스케줄 시간 관리의 효율성을 위해 time format은 [Year,4][Month,2][Date,2][Hour,2][Minute,2][Second,2] 형식으로 총 14자리의 숫자로만 구성
    - DB에 저장되는 스케줄 시간은 **DATE_TIME** 형식으로 저장

4. **공연 스케줄 시간 중첩 방지 기능**
    - PerformanceService에서 스케줄 시간이 정상인지 또는 시간이 중첩되지 않는지 체크
    - 스케줄이 정상이 아니거나 중첩된다면 **PerformanceTimeConflictException** 예외처리
    - 구현:    
        **1. 입력된 스케줄 간의 중첩 체크(checkPerfTimeRequestConflict 메소드)**
            - performanceTimeRequest 리스트를 startTime을 기준으로 정렬
            - startTime과 endTime이 정상적인지 확인(checkCorrectPerfTime)
            - Linear 하게 스케줄이 중첩되지 않는 지 체크
        **2. DB에 저장되어 있는 스케줄과의 중첩 체크(checkPerfTimeWithDB 메소드)**
            - 기존에 존재하는 스케줄을 List로 가져와 startTime을 기준으로 정렬
            `- DB의 스케줄을 가져올 때 같은 홀의 스케줄 중 날짜 범위를 조정하여 체크 알고리즘의 효율 향상`
            `- 공연 시간이 이틀에 걸쳐 진행 될 수 있으므로 입력하려는 스케줄의 시작 시간을 기준으로 앞, 뒤 날짜 모두 SELECT`
            - 정렬된 스케줄을 endTime 기준으로 **이분탐색** 하여 입력하려는 스케줄의 startTime 보다 큰 스케줄 중 가장 작은 스케줄 검색
            - 검색된 스케줄이 입력하려는 스케줄의 시간과 중첩되는지 확인

5. **TEST**
    - 공연 id를 가져오는 것에 한계가 있어 로컬에서 **Postman**으로 테스트 진행
    - 정상적인 요청에 대한 기능은 모두 정상 동작 & 문제 상황에 대한 테스트 수행 및 해결
    - 문제 상황에서의 테스트:
        - **공연 이미지 등록**
            - 파일 크기가 허용 값 보다 초과되는 경우
            - 파일이 이미지가 아닌 경우
        - **공연 스케줄 등록**
            - 스케줄 시간이 정상이 아닌 경우(ex. 끝 시간이 시작 시간보다 빠름)
            - 스케줄이 서로 중첩되는 경우
        - **etc**
            - 관리자 권한이 없는 경우

hanwix2 committed:
[#34] 공연 정보 추가 기능 구현

- 공연(performance), 포스터사진(posterImage) 테이블 생성 sql 작성
- Performance(공연) 관련 Controller, Service, Mapper 생성
- 공연 정보 등록 기능 구현
    - PerformanceRequest 객체로 공연 정보 받아옴
    - insertPerformance 메소드에서 공연 등록 로직 수행
    - 공연 등록 전 공연명과 공연유형이 중복되었으면
      PerformanceAlreadyExistsException 에러
    - 공연 유형은 ShowType Enum으로 처리

f-lab-dev commented on etc/show_ticketing_service.sql:
```diff
@@ -35,4 +35,28 @@ CREATE TABLE venueHall(
+CREATE TABLE performance
+(
+    `id`        INT UNSIGNED    NOT NULL    AUTO_INCREMENT,
+    `title`     VARCHAR(50)     NOT NULL,
```
여기는 검색조건으로 쓰이던데 인덱스를 걸어줘야하지 않을까요?

hanwix2 replied:
넵! 적용하겠습니다!

f-lab-dev reviewed (commented):
코드가 갈수록 깔끔해지네요~

hanwix2 committed:
[#34] 공연(performance) 테이블 index 생성문 추가

hanwix2 committed:
[#34] 공연 정보 관련 DB Table 수정

- 개발 과정을 단순화 하기 위해 포스터 이미지(posterImage) 테이블을 삭제하고 공연(performance) 테이블에 이미지 경로 저장
- 공연 정보 입력 로직에 트랜잭션 적용

hanwix2 committed:
[#34] 공연 포스트 이미지 등록 기능 구현

- MultipartFile을 이용하여 이미지 파일 등록
- form-data로 공연정보와 파일을 동시에 보낼 수 없으므로 기존 공연 등록 API와 분리
- 이미지 등록시 프로젝트 파일의 files/performanceImages 에 저장과 동시에
  해당 공연(performance) 데이터의 imageFilePath 속성에 파일경로 Update
- 파일 저장 시 파일명의 중복을 피하기 위해 UUID를 사용하여 고유 이름 지정

hanwix2 committed:
[#34] 공연 이미지 등록 시 기존 파일 삭제

hanwix2 committed:
[#34] 공연 포스터 등록 시 파일 형식 확인

- 업로드 할 파일이 이미지(jpeg, png, gif) 파일이 아닌 경우
  InvalidFileTypeException 예외 처리

hanwix2 committed:
[#34] 공연 이미지 파일 예외 처리

- 최대 업로드 가능 크기: 10MB
- 허용 크기를 넘어가는 예외(SizeLimitExceededException)에 대해 기존 500 오류를 400 오류로 전환

hanwix2 committed:
[#34] 공연 스케줄 등록 기능 구현

- 이미 존재하는 공연에 대하여 스케줄 입력
- 공연 스케줄은 하나 이상이므로 스케줄은 List로 받음
- PerformanceTimeRequest 객체로 공연 시작, 끝 시간 전달 받음
- 시간 format은 [Year,4][Month,2][Date,2][Hour,2][Minute,2][Second,2] 형식으로 총 14자리의 숫자로만 구성

f-lab-dev commented on src/main/java/com/show/showticketingservice/service/PerformanceService.java:
```diff
@@ -6,13 +6,15 @@
 
     private final PerformanceMapper performanceMapper;
 
+    @Transactional
```
이게 어떤 어노테이션인지 주석 하나 달아줘도 괜찮을 것 같습니다~

hanwix2 replied:
음.. 트랜잭션 기능에 대한 설명일까요? 트랜잭션을 적용한 이유를 말씀하신건가요?
처음 트랜잭션 어노테이션을 사용하는 것이 아니라 다른 트랜잭션 적용한 코드에도 적어줘야하나 고민입니다.

f-lab-dev commented on src/main/java/com/show/showticketingservice/service/FileService.java:
```diff
@@ -0,0 +1,50 @@
+
+@Service
+@RequiredArgsConstructor
+public class FileService {
```
이거는 인터페이스로 추상화해줘도 될 것 같습니다~ 이 클래스는 로컬 저장소에 파일을 업로드하는 기능을 갖고있는 것 같네요~

f-lab-dev commented on src/main/java/com/show/showticketingservice/service/FileService.java:
```diff
@@ -0,0 +1,50 @@
+            return path.toString();
+        } catch (IOException e) {
+            String eMessage = String.format("' %s' 이미지 파일 업로드에 실패하였습니다.", fileName);
+            throw new RuntimeException(eMessage);
```
`e`를 같이 넣어주어서 원인 추적이 가능하게 만들어줘야할 것 같습니다~

f-lab-dev commented on src/main/java/com/show/showticketingservice/tool/util/FileUtil.java:
```diff
@@ -0,0 +1,36 @@
+import java.nio.file.Paths;
+import java.util.UUID;
+
+@Component
```
유틸클래스는 일반적으로 빈으로 등록하지 않고 static 메소드를 제공합니다~

f-lab-dev commented on src/main/resources/application.properties:
```diff
@@ -19,3 +19,9 @@ spring.redis.cache.port=6380
 logging.config=src/main/resources/log/Log4j2.xml
+
+# fileUpload
+    spring.servlet.multipart.enabled=true
```
여기 들여쓰기는 굳이 필요가 없어보이네요~

hanwix2 replied:
_'공연 이미지 파일 예외 처리' (c5a3d32)_ 커밋에서 해당 들여쓰기는 삭제되었습니다!^^

hanwix2 committed:
[#34] 파일 업로드 관련 피드백 반영

- FileUtil 클래스는 빈으로 등록하지 않고 static 메소드 제공
- FileService 추상화
    - LocalFileService: 로컬 저장소 파일 업로드 관련 서비스
- LocalFileService에서 파일 업로드 실패 예외 처리 중 실패 원인 추적을 위한 예외 객체 전달

f-lab-dev commented on src/main/java/com/show/showticketingservice/service/LocalFileService.java:
```diff
@@ -0,0 +1,65 @@
+    private String fileUploadPath;
+
+    @Override
+    public String registerPosterImage(MultipartFile image) {
```
이 클래스는 파일과 관련된 책임을 맡고 있습니다. 파일과 관련된 공통 클래스에서 **포스터** 라는 비즈니스 개념을 알고있으면 어떤 일이 벌어질까요?

hanwix2 replied:
메소드는 포스터 이미지 관련 작업에 한정되고
같은 인터페이스를 상속하는 다른 클래스에서 필요 없는 메소드를 구현해야 하는 불필요한 작업을 수행해야합니다.

hanwix2 committed:
[#34] 공연 스케줄 입력 시 스케줄 중첩 체크

- PerformanceService에서 스케줄 시간 정상인지, 시간이 중첩되지 않는지 체크
- 스케줄이 정상이 아니거나 중첩된다면 PerformanceTimeConflictException 예외처리

1. 입력된 스케줄 간의 중첩 체크(checkPerfTimeRequestConflict 메소드)
    - performanceTimeRequest 리스트를 startTime을 기준으로 정렬
    - startTime과 endTime이 정상적인지 확인(checkCorrectPerfTime)
    - Linear 하게 스케줄이 중첩되지 않는 지 체크

2. DB에 저장되어 있는 스케줄과의 중첩 체크(checkPerfTimeWithDB 메소드)
    - 기존에 존재하는 스케줄을 List로 가져와 startTime을 기준으로 정렬
    - 정렬된 스케줄을 endTime 기준으로 이분탐색 하여 입력하려는 스케줄의
      startTime 보다 큰 스케줄 중 가장 작은 스케줄 검색
    - 검색된 스케줄이 입력하려는 스케줄의 시간과 중첩되는지 확인

hanwix2 committed:
[#34] FileService 메소드명 수정

f-lab-dev commented on src/main/java/com/show/showticketingservice/service/FileService.java:
```diff
@@ -4,9 +4,9 @@
+    String registerImage(MultipartFile image);
 
-    String saveImage(String fileName, MultipartFile image);
+    String saveFile(String fileName, MultipartFile image);
```
`registerImage`와 `saveFile`은 역할이 비슷한 것 같은데 그냥 오버로딩으로 정의하면 되지 않을까요?

hanwix2 replied:
`registerImage`는 파일 저장 전 이미지인지 확인하는 작업을 하는 메소드라 오버라이딩으로 정의하기엔 약간 애매할 수 있는데 
그냥 `saveFile`하나의 메소드로 통합해도 괜찮을까요?

f-lab-dev replied:
아 그렇군요~ 

1. 메소드 이름을 그럼 `파일 저장 전 이미지인지 확인하는 작업`이라는게 명확하게 보여지도록 이름을 바꿔주는게 좋을 것 같고요.
2. 이미지인지 확인하는 작업은 `FileService`의 책임이 맞을까요?

hanwix2 replied:
앗! 제가 잘못 알려드렸습니다. 죄송합니다. 코드를 작성하고 시간이 조금 경과해서 착오가 있었네요 ㅠ

- 이미지 확인 작업은 `checkFileContentType()` 메소드에서 `FileUtil`의 `isImageFile()` 메소드를 호출하여 체크합니다.

- 처음 말씀하신대로 `registerImage`와 `saveFile` 메소드는 역할이 비슷하고, 또 지금 보니 굳이 나눌 필요가 없어 보이네요. 

**하나의 메소드**로 합치겠습니다.

f-lab-dev reviewed (commented):
스케줄쪽은 DB에서 가져오는 날짜 범위 좁혀주는 것 적용해주시면 될 것 같네요~

hanwix2 committed:
[#34] performanceTime 테이블에 hallId 속성 추가

- 공연시간 검사 시 기존에는 같은 공연에 대해서만 검사하던 것을 같은 홀에 대해 검사하는 것으로 변경하기 위함
- 공연 스케줄 입력 시 자동으로 공연 id에 따른 공연 홀 id 입력

hanwix2 committed:
[#34] DB의 공연 스케줄 SELECT 시 날짜 범위 축소

- 공연 스케줄 입력 중 DB와의 스케줄 중첩 검사 시 가져오는 데이터를 최소한으로 하여 성능 향상을 위함
- DB의 스케줄 데이터를 입력하려는 스케줄 시작 시간 기준 전날과 다음날까지의 범위로 SELECT
- 공연 시간이 이틀에 걸쳐 진행 될 수 있기 때문

hanwix2 committed:
[#34] 공연 시간을 최대 24시간으로 제한

tlsgkr7416 commented on src/main/resources/mybatis/mappers/PerformanceTimeMapper.xml:
```diff
@@ -16,10 +16,20 @@
         FROM performanceTime
-        WHERE performanceId = #{performanceId}
+        WHERE hallId = (SELECT hallId FROM performance WHERE id = #{performanceId}) AND
+        <foreach collection="performanceTimeRequests" item="performanceTimeRequest" open="(" close=")" separator="OR">
+            (
+            DATE(startTime)
+            BETWEEN
+            DATE(#{performanceTimeRequest.startTime}) - 1
+            AND
+            DATE(#{performanceTimeRequest.startTime}) + 1
+            )
+        </foreach>
```
여기가 공연 시간이 이틀에 걸쳐서 진행되어 데이터를 갖고 오게 한 건가요? 그런데 스토리가 좀 이해가 안 가는데 공연 시간이 이틀보다 더 이상 진행될 수 있는 거 아닌가요?

hanwix2 replied:
그래서 공연 시간은 최대 24시간으로 지정했습니다.
(사실 그렇게까지 긴 공연은 존재하지 않다고 봐야죠)
이틀에 걸치다는 것은 예를 들어 23시에 시작해서 다음날 01시에 끝나는 공연이 있을수 있다는 것입니다.

tlsgkr7416 replied:
아 이틀에 걸친다는 게 그 경우군요. 감사합니다.

hanwix2 committed:
[#34] FileService의 파일 저장 관련 메소드 통합

- registerImage 와 saveFile 메소드를 saveFile로 통합
- 이유: 기존 registerImage 메소드에선 파일 저장(saveFile) 전 FileUtil을
  활용한 파일명, 저장경로를 확인하는 단순 작업만 처리하기 때문에 파일
  저장 메소드를 구분할 필요가 없음

hanwix2 committed:
Merge branch 'develop' of https://github.com/f-lab-edu/show-ticketing-service into feature/34/jh
